src/Queue/DoctrineQueue.php

```diff
@@ -182,19 +225,162 @@ public function pop(array $options = []): ?JobInterface
             return null;
         }
 
-        // Add job ID to meta data
-        return $this->unserializeJob($row['data'], ['__id__' => $row['id']]);
+        return $row;
+    }
+
+    /**
+     * This generic variant of pop() works like this:
+     * - UPDATE query (limit 1) that sets the job into a new status (STATUS_ALLOCATING) with a unique (!) worker ID
+     * - If the row count of the UPDATE is zero -> no Jobs are ready for execution -> return null.
+     * - If the row count of the UPDATE is greater 1 -> too many Jobs were reserved -> rollback and throw an exception. This should not happen!
+     * - If the row count of the UPDATE is 1 -> we've marked a job for us
+     * - SELECT the job via the unique identifier
+     * - UPDATE it to STATUS_RUNNING
+     * - Return its data.
+     *
+     * It uses 3 queries (UPDATE, SELECT, UPDATE) to pop a job from the queue table without any locking or transaction.
+     *
+     * @param array $options
+     *
+     * @return array|null
+     *
+     * @throws \Exception
+     */
+    protected function popMysql(array $options = []): ?array
+    {
+        $time      = microtime(true);
+        $micro     = sprintf("%06d", ($time - floor($time)) * 1000000);
+        $this->now = new DateTime(
+            date('Y-m-d H:i:s.' . $micro, $time),
+            new DateTimeZone(date_default_timezone_get())
+        );
+
+        if ($this->workerId === null) {
+            $this->workerId = getmypid() . '-' . random_int(0, PHP_INT_MAX);
+        }
+
+        try {
+            $queryBuilder = $this->connection->createQueryBuilder();
+            $queryBuilder
+                ->update($this->options->getTableName(), 'job')
+
+                ->set('job.status = :newStatus')
+                ->setParameter('newStatus', static::STATUS_ALLOCATING)
+
+                ->set('job.workerId = :workerId')
+                ->setParameter('workerId', $this->workerId)
+
+                ->where('job.status = :oldStatus')
+                ->setParameter('oldStatus', static::STATUS_PENDING)
+
+                ->andWhere('job.queue = :queue')
+                ->setParameter('queue', $this->getName())
+
+                ->andWhere('scheduled <= :scheduled')
+                ->setParameter('scheduled', $this->now->format('Y-m-d H:i:s.u'))
+
+                ->addOrderBy('priority', 'ASC')
+                ->addOrderBy('scheduled', 'ASC')
+
+                ->setMaxResults(1);
+
+            $result = $this->connection->executeQuery(
+                $queryBuilder->getSQL(),
+                $queryBuilder->getParameters(),
+                $queryBuilder->getParameterTypes()
+            );
+
+            $rowCount = $result->rowCount();
+
+            if ($rowCount === 0) {
+                // No row was updated -> no job ready for us!
+                return null;
+            }
+
+            if ($rowCount > 1) {
+                // That should not be possible, but what if ...? Reset them...
```

Comment:
Cool that you handled this case. But really: I would be okay with just throwing the exception :D. That would mean less code to test and support, and I feel this is never ever going to happen (that would mean the database engine is fucking it up).